theseus/optimizer/nonlinear/nonlinear_optimizer.py

```diff
@@ -107,6 +107,8 @@ def resolve(key: Union[str, "BackwardMode"]) -> "BackwardMode":
 #                ignoring indices given by `reject_indices`
 #    3. Check convergence
 class NonlinearOptimizer(Optimizer, abc.ABC):
+    _MAX_ALL_REJECT_ATTEMPTS = 3
```

Comment:
Any tradeoff or overhead to consider here? Curious about the choice of `3` attempts. Are these all kept in the compute graph?

Also might be helpful to add in a comment that these max attempts are per iteration and not overall.